src/radiation/moments.cpp

```diff
@@ -271,68 +272,109 @@ TaskStatus ReconstructEdgeStates(T *rc) {
   const int ndim = pmb->pmy_mesh->ndim;
   auto &coords = pmb->coords;
 
-  parthenon::par_for(
-      DEFAULT_LOOP_PATTERN, "RadMoments::Reconstruct", DevExecSpace(), X1DIR,
-      ndim,                 // Loop over directions for reconstruction
-      0, nblock - 1,        // Loop over blocks
-      kb.s - dk, kb.e + dk, // z-loop
-      jb.s - dj, jb.e + dj, // y-loop
-      ib.s - di, ib.e + di, // x-loop
-      KOKKOS_LAMBDA(const int iface, const int b, const int k, const int j, const int i) {
-        ReconstructionIndexer<VariablePack<Real>> ql(ql_base, nrecon, offset, b);
-        ReconstructionIndexer<VariablePack<Real>> qr(qr_base, nrecon, offset, b);
-        // Reconstruct radiation
-        for (int ivar = 0; ivar < nrecon; ++ivar) {
-          PhoebusReconstruction::PiecewiseLinear(iface, ivar, k, j, i, v, ql, qr);
-        }
-        // Reconstruct velocity for radiation
-        for (int ivar = 0; ivar < 3; ++ivar) {
-          PhoebusReconstruction::PiecewiseLinear(iface, ivar, k, j, i, v_vel, ql_v, qr_v);
-        }
+  // TODO(JCD): par_for_outer doesn't have a 4d loop pattern which is needed for blocks
+  parthenon::par_for_outer(
+    DEFAULT_OUTER_LOOP_PATTERN, "RadMoments::Reconstruct", DevExecSpace(),
+    0, 0, 0, nrecon+2, kb.s - dk, kb.e + dk, jb.s - dj, jb.e + dj,
+    KOKKOS_LAMBDA(parthenon::team_mbr_t member, const int n,
+                  const int k, const int j) {
+      const int b = 0; // this will be replaced by the block arg to the lambda
+      ReconstructionIndexer<VariablePack<Real>> ql(ql_base, nrecon, offset, b);
+      ReconstructionIndexer<VariablePack<Real>> qr(qr_base, nrecon, offset, b);
+
+      const VariablePack<Real> &var = (n < nrecon ? v : v_vel);
+      const int var_id = n % nrecon;
+
+      Real *pv = &var(b, var_id, k, j, 0);
+      Real *pvjm1 = &var(b, var_id, k, j-dj, 0);
+      Real *pvjp1 = &var(b, var_id, k, j+dj, 0);
+      Real *pvkm1 = &var(b, var_id, k-dk, j, 0);
+      Real *pvkp1 = &var(b, var_id, k+dk, j, 0);
+      Real *vi_l, *vi_r, *vj_l, *vj_r, *vk_l, *vk_r;
+      if (n < nrecon) {
+        vi_l = &ql(0, n, k, j, 1);
+        vi_r = &qr(0, n, k, j, 0);
+        vj_l = &ql(1%ndim, n, k, j+dj, 0);
+        vj_r = &ql(1%ndim, n, k, j, 0);
+        vk_l = &ql(2%ndim, n, k+dk, j, 0);
+        vk_r = &ql(2%ndim, n, k, j, 0);
+      } else {
+        vi_l = &ql_v(0, var_id, k, j, 1);
+        vi_l = &qr_v(0, var_id, k, j, 0);
+        vj_l = &ql_v(1%ndim, var_id, k, j+dj, 0);
+        vj_l = &qr_v(1%ndim, var_id, k, j, 0);
+        vk_l = &ql_v(2%ndim, var_id, k+dk, j, 0);
+        vk_l = &qr_v(2%ndim, var_id, k, j, 0);
+      }
+
+      // TODO(JCD): do we want to enable other recon methods like weno5?
```

Comment:
In a future pull request we can template this function on reconstruction type as well. There is no reason not to enable other reconstruction methods, although in the past I have seen more M1 instabilities with higher order methods.